File: rx-activity/src/com/hendraanggrian/rx/activity/ActivityResultEmitter.java
package com.hendraanggrian.rx.activity;

import android.support.annotation.NonNull;

import io.reactivex.ObservableEmitter;
import io.reactivex.annotations.Nullable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Cancellable;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
abstract class ActivityResultEmitter<T> implements ObservableEmitter<T> {

    static final int TYPE_OK = 0;
    static final int TYPE_RESULT = 1;

    abstract int getType();

    @NonNull
    static <T> ActivityResultEmitter<T> create(final int type, @NonNull final ObservableEmitter<T> e) {
        return new ActivityResultEmitter<T>() {
            @Override
            int getType() {
                return type;
            }

            @Override
            public void setDisposable(@Nullable Disposable d) {
                e.setDisposable(d);
            }

            @Override
            public void setCancellable(@Nullable Cancellable c) {
                e.setCancellable(c);
            }

            @Override
            public boolean isDisposed() {
                return e.isDisposed();
            }

            @Override
            public ObservableEmitter<T> serialize() {
                return e.serialize();
            }

            @Override
            public void onNext(@io.reactivex.annotations.NonNull T value) {
                e.onNext(value);
            }

            @Override
            public void onError(@io.reactivex.annotations.NonNull Throwable error) {
                e.onError(error);
            }

            @Override
            public void onComplete() {
                e.onComplete();
            }
        };
    }
}

File: rx-activity/src/com/hendraanggrian/rx/activity/ActivityResultEmitter.kt
package com.hendraanggrian.rx.activity

import io.reactivex.ObservableEmitter
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Cancellable

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
internal interface ActivityResultEmitter<T> : ObservableEmitter<T> {

    val type: Int

    companion object {
        val TYPE_OK = 0
        val TYPE_RESULT = 1

        fun <T> getInstance(type: Int, e: ObservableEmitter<T>) = object : ActivityResultEmitter<T> {
            override val type = type

            override fun setDisposable(d: Disposable?) {
                e.setDisposable(d)
            }

            override fun setCancellable(c: Cancellable?) {
                e.setCancellable(c)
            }

            override fun isDisposed() = e.isDisposed

            override fun serialize() = e.serialize()

            override fun onNext(@io.reactivex.annotations.NonNull value: T) {
                e.onNext(value)
            }

            override fun onError(@io.reactivex.annotations.NonNull error: Throwable) {
                e.onError(error)
            }

            override fun onComplete() {
                e.onComplete()
            }
        }
    }
}